api/v1/coupons.js

```diff
@@ -30,6 +32,38 @@ var createCouponForNewUser = function (req, res, next) {
 };
 exports.createCouponForNewUser = createCouponForNewUser;
 
+var getDiscountOrderValueByCouponID = function (req, res, next) {
+  var couponId = req.params.couponID;
+  var username = req.query.username;
+  var orderValue = req.query.orderValue;
+
+  var isBelongToUsers = coupon.isBelongToUsers(couponId, username);
+  var isCouponValid = coupon.isCouponValid(couponId);
+  
+  Promise.join(isBelongToUsers, isCouponValid, function (belong, valid) {
+    if (!belong && valid) return true;
+    return false;
+  }).then(function (ableToUse) {
+    if (ableToUse) {
+        return coupon.getCouponCodesByCouponCode(couponId).then(function (coupons) {
+        return coupons[0];
+      }).then(function (couponObject) {
+        var ruleType = couponObject.couponRule.type;
+        var ruleValue = couponObject.couponRule.value;
+        console.log(ruleType + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+        console.log(ruleValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+        console.log(orderValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+
+        var discountedValue = coupon.getDiscountedValue(ruleType, ruleValue, orderValue);
+        res.statusCode = 201;
```

Comment:
However, this is a style problem, the real source for your error (Error: Can't set headers after they are sent.) is the
+ ` res.send('TODO'); `line 64, is sent outside the promise, hence you have send the res before the promises returned, and then you set the header in the promise return     `res.statusCode = 201;. ` at line 58